sql/parser: Print datum dates/timestamps in apostrophes

a-robinson opened the pull request:
Without this, they get printed out in a format that prevents them from
being properly parsed. This was only hit in a very particular edge case
of using a literal to compare against a DATE column in a view's select
statement.

Fixes #12420.

It's a little concerning that this appears to change the format of keys...

Also, I wonder whether we may want to create the persisted string representation of the query during the construction of the `createViewNode` rather than in its `Start` method to avoid other quirks of the transformations that may be applied to the AST between the two. What do you think, @knz?

Finally, I'm still tracking down the last test failure (`TestCopyRandom`, which is still failing due to how this affects placeholder queries), but sending out now for faster feedback on everything else.

@arjunravinarayan

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12450)
<!-- Reviewable:end -->

a-robinson committed:
sql/parser: Print datum dates/timestamps in apostrophes

Without this, they get printed out in a format that prevents them from
being properly parsed. This was only hit in a very particular edge case
of using a literal to compare against a DATE column in a view's select
statement.

a-robinson commented:
Ok, all tests now passing. `TestCopyRandom` was failing because lib/pq puts apostrophes around timestamp literals, so the bug was just in the test.

knz commented:
Yes making a string rendering earlier seems to be a good idea.

LGTM otherwise

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12450#-:-KZ7ZVhS0d9ksRKWhzCz:b-ikv5g9)*
<!-- Sent from Reviewable.io -->

dt commented:
fwiw, it's only changing the _pretty print_ of the keys as used by DebugVals -- which for a scan node is filled by RowFetcher's prettyEncDatums, which in turn is just implicitly String() ing Datum, thus getting your change to FormatNode.

a-robinson committed:
sql: Format view query string during plan construction

This should expose it to less of the modifications that may be done on a
query between planNode construction and start.

a-robinson commented:
Phew, thanks @dt!

I've added a commit to move the view source query formatting to the planNode's constructor.

arjunravinarayan commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for getting to this!

---

Reviewed 8 of 8 files at r1.
Review status: 8 of 9 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12450#-:-KZ7kBw4CeSeVQTURNyq:b-s5c2iv)*
<!-- Sent from Reviewable.io -->

tamird commented on pkg/sql/parser/eval.go:
```diff
@@ -1930,8 +1930,13 @@ func (expr *CastExpr) Eval(ctx *EvalContext) (Datum, error) {
+		case *DTimestamp, *DTimestampTZ, *DDate:
+			var buf bytes.Buffer
+			d.Format(&buf, FmtBareStrings)
+			fmt.Println(buf.String())
```
what's this?

a-robinson replied:
Ugggghhhh thanks

a-robinson commented on pkg/sql/copy_in_test.go:
```diff
@@ -178,7 +179,9 @@ func TestCopyRandom(t *testing.T) {
-			ds := d.String()
+			var buf bytes.Buffer
+			d.Format(&buf, parser.FmtBareStrings)
+			ds := buf.String()
```
Thanks, added that onto #12454
